micrometer-commons/src/main/java/io/micrometer/common/Tag.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2017 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.common;
+
+/**
+ * Key/value pair representing a dimension of a meter used to classify and drill into measurements.
+ *
+ * @author Jon Schneider
+ */
+public interface Tag<T extends Tag<T>> extends Comparable<T> {
+    String getKey();
+
+    String getValue();
+
+    static Tag of(String key, String value) {
+        return new ImmutableTag(key, value);
+    }
+
+    @Override
+    default int compareTo(T o) {
```

Comment:
Is this why we need the generic param?
If this is the only reason, instead of doing this, can we use a `Comparator` where we would use this?
I think the only place is
```
Arrays.sort(this.tags);
```
that can be replaced to:
```
Arrays.sort(this.tags, Comparator.comparing(Tag::getKey));
```
And later to an universal `Comparator` that can compare both tags.

Reply:
Actually, it seems that this can be at all removed - I mean the generic type.